lib/prosopite.rb

```diff
@@ -281,14 +281,20 @@ def subscribe
         sql, name = data[:sql], data[:name]
 
         if scan? && name != "SCHEMA" && sql.include?('SELECT') && data[:cached].nil? && !ignore_query?(sql)
-          query_caller = caller
-          location_key = Digest::SHA256.hexdigest(query_caller.hash.to_s)
+          query_caller = caller_locations
+          # Calculate the location key with as few allocations as possible
+          location_key = [].tap { |array|
+            query_caller.each { |loc|
+              array << loc.path
+              array << loc.lineno
+            }
+          }.hash
```

Comment:
As we just need a consistent and unique hash key for this location, we don't need the SHA256 digest. Unfortunately,  the [`Thread::Backtrace::Location`](https://ruby-doc.org/core-3.1.1/Thread/Backtrace/Location.html) we get from `caller_locations` don't return consistent values for `hash`, so we have to compute the hash manually from the paths and line numbers here.

As you can see from the benchmark in the PR description, adding these values to a single array before calling `hash` on that array is a lot faster than `map`'ing the values to another array.

Reply:
I also wondered if we needed to add `loc.label` to the array here, as that clarifies what method and/or block the call came from (e.g. `block (2 levels) in Pry#eval` vs `Pry#eval`) but I couldn't find a test case where omitting it caused any issues, and including it makes the logic substantially slower. If omitting it does cause any issues, it would only be that separate queries coming from the same call depth on the same line would be flagged as potential N+1s incorrectly, but that seems like an unlikely edge case and impact of the false positive would likely be minor.

Reply:
Either way, I added a test case for asserting that two separate queries from different call depths on the same line don't fire an error.